confmap/resolver.go

```diff
@@ -308,6 +286,90 @@ func (mr *Resolver) expandValue(ctx context.Context, value any) (any, bool, erro
 	return value, false, nil
 }
 
+// findURI will find the URI corresponding to the first closing bracket in input.
+// findURI is only called when input contains a }.
+func (mr *Resolver) findURI(input string) (string, bool) {
+	closeIndex := closeIndex(input)
+	openIndex := openIndex(input[:closeIndex+1])
+	// Should not expand because there is a missing ${.
+	if openIndex < 0 {
+		return "", false
+	}
+
+	URI := input[openIndex : closeIndex+1]
+
+	// Should not expand. This is expanded in the expandconverter.
+	if !strings.Contains(URI, ":") {
+		return "", false
+	}
+
+	return URI, true
+}
+
+func closeIndex(s string) int {
+	return strings.Index(s, "}")
+}
+
+func openIndex(s string) int {
+	return strings.LastIndex(s, "${")
+}
+
+func (mr *Resolver) expandURIs(ctx context.Context, input string) (string, bool, error) {
+	var err error
+	var changed bool
+
+	URI, expand := mr.findURI(input)
+	if expand {
+		input, changed, err = mr.expandableURI(ctx, input, URI)
+	}
+	// Check if other URIs are expandable.
+	if !expand {
+		input, changed, err = mr.nonExpandableURI(ctx, input)
+	}
+	return input, changed, err
+}
+
+func (mr *Resolver) expandableURI(ctx context.Context, input string, uri string) (string, bool, error) {
+	repl, changed, err := mr.expandURI(ctx, uri)
+	return strings.ReplaceAll(input, uri, repl), changed, err
+}
+
+// nonExpandableURI finds the next expandable URI in input and expands it.
+func (mr *Resolver) nonExpandableURI(ctx context.Context, input string) (string, bool, error) {
+	var err error
+	var expandedRemaining string
+	var changed bool
+
+	noExpand := input[:closeIndex(input)+1]
+	remaining := input[closeIndex(input)+1:]
+
+	if strings.Contains(remaining, "}") {
+		expandedRemaining, changed, err = mr.expandURIs(ctx, remaining)
+		return noExpand + expandedRemaining, changed, err
+	}
+	return input, changed, err
+}
+
+func (mr *Resolver) expandURI(ctx context.Context, value string) (string, bool, error) {
```

Comment:
Can we document the newly added functions a bit more? I find the existence of `expandURI` and `expandURIs` functions that accept and return exactly the same parameters confusing. Perhaps rename to make the purpose clearer?

Generally I find it hard to reason about how this new set of functions call recursively and how the recursion is supposed to be finite. Documentation may help with this too.